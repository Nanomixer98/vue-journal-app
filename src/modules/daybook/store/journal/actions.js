import journalApi from "@/api/journalApi"
// export const myAction = async ({ commit }) => {
// }

export const loadEntries = async ({ commit }) => {
    const { data } = await journalApi.get('/entries.json')
    if( !data ) {
        commit('setEntries', [])
        return
    }
    const entries = []
    for (const id of Object.keys(data)) {
        entries.push({
            id,
            ...data[id]
        })
    }
    
    commit('setEntries', entries)
}

// export const updateEntry = async ({ commit }, { id, ...dataToSave}) => {
//     const { data } = await journalApi.put(`/entries/${id}.json`, dataToSave)
//     commit('updateEntry', {id, ...data})
// }

export const updateEntry = async ({ commit }, entry) => {
    const { date, picture, text } = entry
    const dataToSave = { date, picture, text }
        
    // await journalApi.put(`/entries/${entry.id}.json`, dataToSave)
    
    commit('updateEntry', { ...dataToSave, id: entry.id })
}

export const createEntry = async ({ commit }, newEntry) => {
    const { data } = await journalApi.post(`/entries.json`, newEntry)
    const id = data.name
    commit('addEntry', {id, ...newEntry})
    return id
}

export const deleteEntry = async ({ commit }, id) => {
    await journalApi.delete(`/entries/${id}.json`)
    commit('deleteEntry', id)
}