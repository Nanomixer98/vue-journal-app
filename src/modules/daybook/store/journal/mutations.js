// export const myMutation = ( state ) => {
// }

export const setEntries = ( state, entries ) => {
    state.entries = [...state.entries, ...entries]
    state.isLoading = false
}

export const updateEntry = (state, updatedEntry) => {
    const entryIdx = state.entries.findIndex(entry => entry.id == updatedEntry.id)
    state.entries[entryIdx] = updatedEntry
}

export const addEntry = ( state, newEntry ) => {
   state.entries = [newEntry, ...state.entries]
}

export const deleteEntry = ( state, deltedEntryId ) => {
    state.entries = state.entries.filter(( entry ) => entry.id !== deltedEntryId);
}