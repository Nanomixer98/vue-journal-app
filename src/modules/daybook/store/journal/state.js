export default () => ({
    isLoading: true,
    entries: [
        {
            id: new Date().getTime(),
            date: new Date().toDateString(),
            text: 'Lorem ipsum dolor sit amet consectetur adipisicing elit. Beatae, voluptatum consequuntur. Perferendis blanditiis praesentium perspiciatis.',
            picture: null
        },
        {
            id: new Date().getTime() + 100,
            date: new Date().toDateString(),
            text: 'Mollit id aute aute dolore commodo elit sint sit.',
            picture: null
        },
        {
            id: new Date().getTime() + 200,
            date: new Date().toDateString(),
            text: 'Incididunt elit aute officia adipisicing cupidatat enim amet cillum mollit anim ullamco veniam labore.',
            picture: null
        },
    ]
})