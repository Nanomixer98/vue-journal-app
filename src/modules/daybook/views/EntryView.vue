<template>
  <template v-if="entry">
    <div class="entry-title d-flex justify-content-between p-2">
      <div>
        <span class="text-success fs-3 fw-bold">{{ day }}</span>
        <span class="mx-1 fs-3">{{ month }}</span>
        <span class="mx-2 fs-4 fw-ligth">{{ yearDay }}</span>
      </div>

      <div>
        <input
          type="file"
          @change="onSelectedImage"
          ref="imageSelector"
          v-show="false"
          accept="imaga/png, image/jpeg, image/png"
        />

        <button
          v-if="entry.id"
          class="btn btn-danger mx-2"
          @click="onDeleteEntry"
        >
          Borrar
          <i class="fa fa-trash-alt"></i>
        </button>
        <button class="btn btn-primary" @click="onSelectImage">
          Subir foto
          <i class="fa fa-upload"></i>
        </button>
      </div>
    </div>

    <hr />

    <div class="d-flex flex-column px-3 h-75">
      <textarea placeholder="¿Qué sucedió hoy?" v-model="entry.text"></textarea>
    </div>

    <img
      v-if="entry.picture && !localImage"
      :src="entry.picture"
      alt="entry-pic"
      class="img-thumbnail"
    />
    <img
      v-if="localImage"
      :src="localImage"
      alt="entry-pic"
      class="img-thumbnail"
    />
  </template>

  <Fab icon-name="fa-save" @on:click="saveEntry" />
</template>

<script>
import { defineAsyncComponent } from "vue";
import { mapGetters, mapActions } from "vuex";
import Swal from "sweetalert2";
import getDayMonthYear from "../helpers/getDayMonthYear";
import uploadImage from "../helpers/uploadImage";

export default {
  name: "EntryView",

  props: {
    id: {
      type: String,
      required: true,
    },
  },

  data() {
    return {
      entry: null,
      localImage: null,
      file: null,
    };
  },

  components: {
    Fab: defineAsyncComponent(() => import("../components/FabComponent.vue")),
  },

  computed: {
    ...mapGetters("journal", ["getEntryById"]),
    day() {
      const { day } = getDayMonthYear(this.entry.date);
      return day;
    },
    month() {
      const { month } = getDayMonthYear(this.entry.date);
      return month;
    },
    yearDay() {
      const { yearDay } = getDayMonthYear(this.entry.date);
      return yearDay;
    },
  },

  methods: {
    ...mapActions("journal", ["updateEntry", "createEntry", "deleteEntry"]),

    loadEntry() {
      let entry;
      this.localImage = null;
      if (this.id == "new") {
        entry = {
          text: "",
          date: new Date().getTime(),
        };
      } else {
        entry = this.getEntryById(this.id);
        if (!entry) return this.$router.push({ name: "no-entry" });
      }

      this.entry = entry;
    },

    async saveEntry() {
      Swal.fire({
        title: "Espere por favor",
        allowOutsideClick: false,
      });
      Swal.showLoading();

      const picture = await uploadImage(this.file);
      this.entry.picture = picture;

      if (this.entry.id) {
        await this.updateEntry(this.entry);
      } else {
        const id = await this.createEntry(this.entry);
        this.$router.push({ name: "entry", params: { id } });
      }

      this.file = null;
      Swal.fire("Guardado", "Entrada registrada con éxito", "success");
    },

    async onDeleteEntry() {
      const { isConfirmed } = await Swal.fire({
        title: "¿Está seguro",
        text: "Una vez borrado, no se puede recuperar",
        showDenyButton: true,
        confirmButtonText: "Si, estoy seguro",
      });

      if (isConfirmed) {
        Swal.fire({
          title: "Espere por favor",
          allowOutsideClick: false,
        });
        Swal.showLoading();
        await this.deleteEntry(this.entry.id);
        this.$router.push({ name: "no-entry" });

        Swal.fire("Eliminado", "", "success");
      }
    },

    onSelectedImage(event) {
      const file = event.target.files[0];
      if (!file) {
        this.localImage = null;
        this.file = null;
        return;
      }

      this.file = file;

      const fr = new FileReader();
      fr.onload = () => (this.localImage = fr.result);
      fr.readAsDataURL(file);
    },

    onSelectImage() {
      this.$refs["imageSelector"].click();
    },
  },

  created() {
    this.loadEntry();
  },

  watch: {
    id() {
      this.loadEntry();
    },
  },
};
</script>

<style lang="scss" scoped>
textarea {
  border: none;
  font-size: 20px;
  height: 100%;

  &:focus {
    outline: none;
  }
}

img {
  bottom: 150px;
  box-shadow: 0px 5px 10px rgba($color: #000000, $alpha: 0.2);
  position: fixed;
  right: 20px;
  width: 200px;
}
</style>
