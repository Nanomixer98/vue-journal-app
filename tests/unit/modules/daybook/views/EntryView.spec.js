import { createStore } from "vuex"
import { shallowMount } from "@vue/test-utils";
import Swal from 'sweetalert2';

import journal from '@/modules/daybook/store/journal';
import EntryView from '@/modules/daybook/views/EntryView'
import { journalState } from "@/../tests/unit/mock-data/test-journal-state";

const createVuexStore = ( initalState ) => 
    createStore({
        modules: {
            journal: {
                ...journal,
                state: { ...initalState }
            }
        }
    })

jest.mock('sweetalert2', () => ({
    fire: jest.fn(),
    showLoading: jest.fn(),
    close: jest.fn()
}))

describe('Pruebas en el EntryView', () => {
    
    const mockRouter = {
        push: jest.fn()
    }

    const store = createVuexStore(journalState)
    store.dispatch = jest.fn()

    let wrapper

    beforeEach(() => {
        jest.clearAllMocks()
        wrapper = shallowMount(EntryView, {
            props: {
                id: '-NBn1BMKEse6zUQBPzx1'
            },
            global: {
                mocks: {
                    $router: mockRouter
                },
                plugins: [ store ]
            }
        })
    })

    it('debe de sacar al usuario porque el id no existe', () => {
        const wrapper = shallowMount(EntryView, {
            props: {
                id: 'This ID doesnt exisits'
            },
            global: {
                mocks: {
                    $router: mockRouter
                },
                plugins: [ store ]
            }
        })

        expect(mockRouter.push).toHaveBeenCalledWith({ name: "no-entry" })
    });

    it('debe de mostrar la entrada correctamente', () => {
        expect(wrapper.html()).toMatchSnapshot();
        expect(mockRouter.push).not.toHaveBeenCalledWith()
    });

    it('debe de borrar la entrada y salir ', (done) => {
        Swal.fire.mockReturnValueOnce( Promise.resolve({ isConfirmed: true }))
        wrapper.find('.btn-danger').trigger('click')

        expect(Swal.fire).toHaveBeenCalledWith({
            title: "¿Está seguro",
            text: "Una vez borrado, no se puede recuperar",
            showDenyButton: true,
            confirmButtonText: "Si, estoy seguro",
          })

          setTimeout(() => {
            expect(store.dispatch).toHaveBeenCalledWith("journal/deleteEntry", "-NBn1BMKEse6zUQBPzx1")
              expect(mockRouter.push).toHaveBeenCalled()
              done()
          }, 1);
    });

});