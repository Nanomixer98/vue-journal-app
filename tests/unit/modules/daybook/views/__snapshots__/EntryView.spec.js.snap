// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas en el EntryView debe de mostrar la entrada correctamente 1`] = `
<div class="entry-title d-flex justify-content-between p-2">
  <div><span class="text-success fs-3 fw-bold">12</span><span class="mx-1 fs-3">Septiembre</span><span class="mx-2 fs-4 fw-ligth">2022, Lunes</span></div>
  <div><input type="file" accept="imaga/png, image/jpeg, image/png" style="display: none;"><button class="btn btn-danger mx-2"> Borrar <i class="fa fa-trash-alt"></i></button><button class="btn btn-primary"> Subir foto <i class="fa fa-upload"></i></button></div>
</div>
<hr>
<div class="d-flex flex-column px-3 h-75"><textarea placeholder="¿Qué sucedió hoy?"></textarea></div>
<!--v-if-->
<!--v-if-->
<fab-stub icon-name="fa-save"></fab-stub>
`;
